// magnet/build/build/webpack.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getDefaultWebpackConfig = getDefaultWebpackConfig;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _progressBarWebpackPlugin = require('progress-bar-webpack-plugin');

var _progressBarWebpackPlugin2 = _interopRequireDefault(_progressBarWebpackPlugin);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @param {Magnet} magnet
 * @return {Object}
 */
function getDefaultWebpackConfig(magnet) {
  return {
    output: {
      path: _path2.default.join(magnet.getDirectory(), '.magnet'),
      filename: '[name]'
    },
    plugins: [new _progressBarWebpackPlugin2.default({
      summary: false
    })],
    entry: {},
    resolveLoader: {
      modules: [_path2.default.join(__dirname, '../../node_modules'), 'node_modules']
    },
    module: {
      loaders: []
    },
    node: {
      console: true,
      assert: 'empty',
      async_hooks: 'empty',
      buffer: 'empty',
      child_process: 'empty',
      cluster: 'empty',
      crypto: 'empty',
      dgram: 'empty',
      dns: 'empty',
      domain: 'empty',
      events: 'empty',
      fs: 'empty',
      http: 'empty',
      https: 'empty',
      net: 'empty',
      os: 'empty',
      path: 'empty',
      punycode: 'empty',
      querystring: 'empty',
      readline: 'empty',
      repl: 'empty',
      stream: 'empty',
      string_decoder: 'empty',
      tls: 'empty',
      tty: 'empty',
      url: 'empty',
      util: 'empty',
      v8: 'empty',
      vm: 'empty',
      zlib: 'empty'
    }
  };
}
